### ISSUE TITLE
Networking API helpers

### ISSUE THREAD

**Issue by killua-eu** [lxc/lxd#2150]
After playing around with several containers and figuring out the basic ins and outs of routing and proxying, I could definitely use some help from lxd itself. Follows some suggestions:
- on **lxc launch** allow selecting a specific ip, istead of getting a random one. i.e. `lxc launch ubuntu-daily:16.04 u1 -4 10.12.1.1`
- expose hostnames/containernames to containers, so that intercontainer communication is namebased, not ip based.
- have the daemon monitor which ports on a container have a listener, i.e. `lxc ports nginx` would return 80,443 (esp. interesting with the point below)
- allow to hook into events i.e. if a
  -  new container is launched and a specific parameter is passed/set, launch a script 
  - a container is started/stopped do something else 

Basically anything that would allow me to automate reconfiguring iptables and a proxy container (in my case nginx) routing requests to the correct container would be really helpful.

Should I be suggesting something that was already implemented and I just didnt RTFM well enough, point that out for me, pretty please. You will make my life with lxd more lovable.

p.s.: This ticket is very loosely inspired on how https://github.com/jwilder/nginx-proxy works

**Comment by stgraber** [comment]
Some of that will likely be covered with the new network API we're planning, we'll look at this issue again once that's done.

**Comment by killua-eu** [comment]
great! looking forward to the api & thanks for all the work sofar!

**Comment by stgraber** [comment]
Current LXD runs its own dnsmasq process which manages DNS (there are three configurable mode) so that containers can resolve each other rather than rely on IPs. You can also set static IP allocations through the ipv4.address and ipv6.address properties of the nic device now.

Hooking into events can be done to some extent through the /events API though it may not exactly be the most user friendly interface for this. We may consider developing something better, possibly a different class of events which aren't arbitrary strings of text but are typically very much opposed to shelling out to random binaries from LXD itself as it makes for a very random, hard to debug experience and makes things differ between hosts during migration.

For the ports, it's something we could return per-interface in the /state API but not in a way that'd be that useful to you. It would report ports bound for every interface with what IP they're bound for but from the outside you have no way to know what interface is actually exposed and which isn't.

A container can absolutely do:

    ip link set eth0 name blah
    ip link add type dummy name eth0

At which point things bound to "eth0" are actually container-only (loopback-like) and things bound to "blah" are exposed to the outside.


Anyway, since the bulk of this has been fixed with the network API. I'm closing the issue. Feel free to open feature requests for any feature that you'd still need and we can consider them then.
